### ISSUE TITLE
Irreducible control flow graphs

### ISSUE THREAD

**Issue by petemud** [boogie-org/boogie#921]
```boogie
procedure unsound() {
    var x: int;
    assume(x == 0);
    if (true) {
        goto Inside;
    }
    while (x < 10000) {
        Inside: x := x + 1;
    }
    assert(x == -1);
}
```

```posh
$ boogie unsound.bpl /extractLoops
Boogie program verifier finished with 1 verified, 0 errors
```

## History of commits
1. d8b239c0c3bd5d33f683eafb735aef41dac28760 @shazqadeer `added code to handle irreducible graphs`
2. 28d46f807724f88d77b113200cfb769cb76fa28b @shazqadeer reverted it on the same day
3. 64b57a073afe03ab57d99c96ef6994b77f786d0e @akashlal `Support for irreducible graphs (with extractLoops)`
4. b6796ce965fd1a973e7df19a3717be9187314053 @shazqadeer finally purged his initial code away from repo

So now the only surviving support for irreducible flow graphs I found is undocumented `/extractLoops` option

## Problem with `/extractLoops`
Let's say, we have the canonical nonreducible flow graph
```mermaid
graph LR
    A --> B & C
    B --> C
    C --> B
```
```boogie
procedure nonreducible() {
    A: goto B, C;
    B: goto C;
    C: goto B;
}
```
---
According to the Dragon Book _(9.7.6 Handling Nonreducible Flow Graphs)_ this flow graph should be transformed into one of
```mermaid
graph LR
    A --> B' & C
    B & B' --> C
    C --> B
```
```mermaid
graph LR
    A --> B & C'
    B --> C
    C & C' --> B
```
```boogie
procedure reducible1() {
    A: goto B', C;
    B': goto C;
    B: goto C;
    C: goto B;
}

procedure reducible2() {
    A: goto B, C';
    C': goto B;
    B: goto C;
    C: goto B;
}
```
---
Meanwhile `/extractLoops` just unrolls loop and produces control flow which is not equivalent to original
```mermaid
graph LR
    A --> B & C
    B --> C
    C --> B'1
    B'1 --> C'1
    C'1 --> B'2
    B'2 --> ...
```
```boogie
procedure reducible() {
    A: goto B, C;
    B: goto C;
    C: goto B'1;
    B'1: goto C'1;
    C'1: goto B'2;
    B'2: goto C'2;
    ...
}
```

## Real-world example
https://github.com/open-s4c/libvsync/blob/main/include/vsync/atomic/internal/arm64.h#L551-L564
```mermaid
graph TB
    A[ldx oldv, a
cmp oldv, cmpv]
    B[b.eq oldv]
    W[wfe]
    C[ldx oldv, a
cmp oldv, cmpv]
    D[b.neq oldv]
    E[add new, oldv, v
stx tmp, new, a]
    F[cbnz tmp]
    A --> B
    B --> W & E
    W --> C
    C --> D
    D --> W & E
    E --> F
    F --> C
```
```boogie
procedure await_eq_add() {
    A: goto B;
    B: goto W, E;
    W: goto C;
    C: goto D;
    D: goto W, E;
    E: goto F;
    F: goto C, G;
    G: return;
}
```

## Conclusions
Are there any unsolvable problems with properly handling irreducible flow graphs, that it was removed by @shazqadeer 28d46f807724f88d77b113200cfb769cb76fa28b? Can it be brought back or are there other ways of dealing with this issue in Boogie?

## Quotes

> We use the standard techniques for converting an irreducible graph into an equivalent, although possibly far larger, reducible graph. We are looking into ways to deal with irreducible graphs that avoid this problem, but so far it has not been an issue.
> [Barnett, M., Leino, K.R.M. (2005). Weakest-precondition of unstructured programs.](https://www.microsoft.com/en-us/research/publication/weakest-precondition-of-unstructured-programs/)

> ```
> if (!g.Reducible)
> {
>     throw new VCGenException("Irreducible flow graphs are unsupported.");
> }
> ```
> [Barnett, M. (2009). Initial set of files.](https://github.com/boogie-org/boogie/blob/ce1c2de044c91624370411e23acab13b0381949b/Source/VCGeneration/VC.ssc#L1797)

**Comment by petemud** [comment]
@shazqadeer Is there any hope this gets implemented?

**Comment by petemud** [comment]
```boogie
procedure unsound() {
	var x: int;
	assume(x == 0);
	if (true) {
		goto Inside;
	}
	while (x < 10000) {
		Inside: x := x + 1;
	}
	assert(x == -1);
}
```

```posh
$ boogie unsound.bpl /extractLoops
Boogie program verifier finished with 1 verified, 0 errors
```

**Comment by shazqadeer** [comment]
Thanks for bringing this issue to our attention. I do not remember now why support for them was dropped. Irreducible graphs have not surfaced before in Boogie applications so we never considered adding support for them.  Do you have any concrete application where you need them?

**Comment by petemud** [comment]
> Do you have any concrete application?
Yes. See "Real-world example" from above. I'm using boogie to prove properties of libvsync atomics, which are implemented in ARM assembly

**Comment by shazqadeer** [comment]
ok, I will look into it. But it will take me some time. Do you think you can make progress by transforming to irreducible CFGs on your side? I assume you have a toolchain for generating Boogie and you can do the transformation in that toolchain.

**Comment by petemud** [comment]
I don't actually have any toolchain. The project is small. Part of it is hand-written Boogie model for ARM. And part is a parser from assembly

---

Irreducibility shouldn't be a problem really. The transformations Boogie makes are:

1. Split `assert` at the beginning of basic block into `assert` at the end of all direct predecessors and `assume` instead of `assert` in current block
2. Cut back edges by `havoc`-ing variables that change in loop

Both of these don't require irreducibility. Irreducibility just ensures that there is only one loop header, so you can find where to apply 1 and 2. In irreducible graph - with more than one loop entry - you can just use the one that comes first in code as loop header

**Comment by shazqadeer** [comment]
I studied this issue more thoroughly. It appears to me that there are two separate issues going on:

- The behavior of the option ```/extractLoops``` is unpredictable if an irreducible loop is encountered. The loop extraction procedure (extraction to a recursive procedure) works only for irreducible loops. Today, this option silently unrolls loops to a depth supplied by another option ```RecursionBound```. I think it is better to decouple the behavior of ```/extractLoops``` from ```RecursionBound``` and instead tie it to the option ```LoopUnrollCount``` which can be set from the command line. The default of ```LoopUnrollCount``` is -1 which means no unrolling and which could be used to propagate the IrreducibleGraph exception during loop extraction.

- Boogie does not support irreducible graphs and we have a user who needs this feature. So we should try to add this feature.

**Comment by petemud** [comment]
@gauravpartha You are absolutely correct. I was wrong about cutting back edges in irreducible graph, because it also cuts a forward edge from corresponding reducible graph.

```mermaid
graph LR
    A["`*A:*`"]
    B["`*B:* x := 1`"]
    C["`*C:*`"]
    D["`*D:* **assert** x == 1`"]
    E["`*E:*`"]
    F["`*F:*`"]
    A --> B & E
    B --> C
    C --> D
    D --> E
    %%E -. blocks between E to C .-> C
    E --> F
    F -.-> C
```
```mermaid
graph LR
    A["`*A:*`"]
    B["`*B:* x := 1`"]
    C["`*C:*`"]
    D["`*D:* **assert** x == 1`"]
    E["`*E:*`"]; E'["`*E':*`"]
    F["`*F:*`"]; F'["`*F':*`"]
    A --> B & E'
    B --> C
    C --> D
    D --> E
    %%E -.blocks between E to C.-> C
    %%E' -.blocks between E to C.-> C
    E --> F
    F -.-> C
    E' --> F'
    F' -.-> C
```
---
I still think - not sure if it's advantageous though - that this issue can be worked around without copying all of the blocks by:
1. Not `havoc`ing anything
2. Only relying on pushing `assert`s from loop head to its predecessors, i.e. only using invariants.
_For reducible loops automatically `assert`ing variables that don't change is as easy as `havoc`ing those that change_
3. Cutting ALL of the edges pointing to loop head
```mermaid
graph LR
    A["`*A:*`"]
    B["`*B:* x := 1`"]
    C["`*C:*`"]
    D["`*D:* **assert** x == 1`"]
    E["`*E:*`"]
    F["`*F:*`"]
    A --> B & E
    B -.-> C
    C --> D
    D --> E
    %%E -.blocks between E to C.-> C
    %%E' -.blocks between E to C.-> C
    E --> F
    F -.-> C
```
4. _Perhaps make some experimental tool that does this_
<details>
<summary>

_An example of verification condition in the style of [WP of unstructured programs](https://dl.acm.org/doi/10.1145/1108768.1108813). Which fails. As it should_
</summary>

```smt2
(declare-const x Int)

(declare-const A_ok bool)
(declare-const B_ok bool)
(declare-const C_ok bool)
(declare-const D_ok bool)
(declare-const E_ok bool)
(declare-const F_ok bool)

(define-fun A_be () Bool
  (= A_ok
    (and B_ok E_ok)
  )
)
(define-fun B_be () Bool
  (= B_ok
    (=> (= x 1) true)
  )
)
(define-fun C_be () Bool
  (= C_ok D_ok)
)
(define-fun D_be () Bool
  (= D_ok
    (and (= x 1) E_ok)
  )
)
(define-fun E_be () Bool
  (= E_ok F_ok)
)
(define-fun F_be () Bool
  (= F_ok true)
)

(assert
  (not
    (=>
      (and
        A_be
        B_be
        C_be
        D_be
        E_be
        F_be
      )
      (and
        A_ok
        C_ok
      )
    )
  )
)

(check-sat)
(get-model)
```
</details>

**Comment by shazqadeer** [comment]
Even if an automatic method for converting an irreducible CFG to a reducible CFG could be implemented, I don't understand how it would address the problem of specifying the appropriate loop invariant needed for verifying the loop. For well-structured (or reducible) loops, it is clear to the programmer where to put the loop invariant and this information is exploited by Boogie to generate the appropriate verification conditions.

Any suggestions?

**Comment by petemud** [comment]
Here are a couple of suggestions:
1. Use as loop header the block that comes first in program code
2. Use block that has `assert`s; fail if both loop entries have them
3. Reuse `invariant` keyword. This keyword would only be allowed at the beginning of loop header and have the same meaning as `assert`. Again, if both loop entries have `invariant`, then fail

**Comment by JonasOberhauser** [comment]
I think every loop entry point needs to have invariants.

**Comment by shazqadeer** [comment]
Is anybody volunteering to implement this feature in Boogie?

**Comment by shazqadeer** [comment]
Completed via pull request #1032 .
